nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+import nnvm
+from nnvm.compiler import graph_util
+from nnvm.compiler.graph_attr import TCODE_TO_DTYPE
+from .config import ctx_list
+
+def graph_to_function(graph, target, ctx):
+    """Convert a graph to a function taking a keyword args and returning a list of results
+    (both args and results are numpy arrays).
+
+    Example::
+
+        fun = graph_to_function(graph, llvm, cpu(0))
+        [res1, res2] = fun(x=np.zeros((1,2)), y=np.zeros((1,)))
+
+    Parameters
+    ----------
+    graph : nnvm.graph.Graph
+        A graph we want to convert to a function.
+
+    target : str or :any:`tvm.target.Target`, optional
+        The build target
+
+    ctx : TVMContext
+        The context to deploy the module.
+
+    Returns
+    -------
+    function : Callable[..., List[numpy.ndarray]]
+    """
+
+    shapes = graph.json_attr('shape')
+    dtypes = graph.json_attr('dtype')
+
+    if shapes is None or dtypes is None:
+        graph = graph.apply('InferShape').apply('InferType')
+        shapes = graph.json_attr('shape')
+        dtypes = graph.json_attr('dtype')
+
+    dtypes = [TCODE_TO_DTYPE[dtype] for dtype in dtypes]
+
+    ishapes = {x: shapes[graph.index.entry_id(x)] for x in graph.index.input_names}
+    idtypes = {x: dtypes[graph.index.entry_id(x)] for x in graph.index.input_names}
+
+    compute_graph, lib, _ = nnvm.compiler.build(graph, target, shape=ishapes, dtype=idtypes)
+    module = graph_runtime.create(compute_graph, lib, ctx)
+
+    shapes = compute_graph.json_attr('shape')
+    dtypes = [TCODE_TO_DTYPE[dtype] for dtype in compute_graph.json_attr('dtype')]
+
+    def run(**kwargs):
+        module.run(**kwargs)
+        res = []
+        for i, out_entry in enumerate(compute_graph.index.output_entries):
+            res.append(
+                module.get_output(
+                    i, tvm.nd.empty(shapes[out_entry[0]], dtypes[out_entry[0]])).asnumpy())
+        return res
+
+    return run
+
+
+def check_function(symbol, grad_input_vars=None, np_forward=None, np_backward=None,
+                   shape=None, dtype=None, in_range=None,
+                   exclude_targets=None, only_targets=None,
+                   numerical_grads='if_possible', delta=1e-2,
+                   atol=1e-5, rtol=1e-5, ng_atol=1e-2, ng_rtol=1e-2,
+                   ng_max_error=1e+3, ng_max_discarded_frac=0.1,
+                   dump_graph=False):
+    """Compute the function and/or its gradients on a random input and raise
+    an exception if the result doesn't match the reference implementation.
+
+    Parameters
+    ----------
+    symbol : nnvm.Symbol
+        A symbol representing the output.
+
+    grad_input_vars : List[nnvm.Symbol or str or (str, Tuple[int])], optional
+        A list of variables with respect to which the gradients will be computed.
+        None (default) means that all input variables will be used. May be a
+        list of pairs `(var_name, shape)`.
+
+    np_forward : Callable[..., List[numpy.ndarray]], optional
```

Comment:
Suggest to give different naming for np_forward : Need not be numpy always, could be an implementation in any of the frontends we have.

Also receive and pass a flexible parameters 'attr' (dictionary object with default value None) for this function which can carry the implementation specific data like session info ...etc.